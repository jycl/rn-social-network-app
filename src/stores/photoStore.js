import { observable, computed, toJS, action } from "mobx";
import { getAlbumsForUser, getPhotosForAlbum } from "../services/APIService";

/**
 * PhotoStore is an MobX store that manages the state values of components related
 * to photo albums and photo display.
 *
 * State params:
 * @param {Array} rawAlbumList Array of posts for user retrieved from backend API
 * @param {Array} albumList list of albums that has been converted from MobX object
 */
class PhotoStore {
  @observable
  rawAlbumList = [];

  @observable
  currentAlbum = null;

  @observable
  currentPhoto = null;

  @observable
  albumPhotoListMapping = new Map();

  /**
   * Retrieve user list from backend and save to store
   */
  @action
  loadPhotoAlbums = userId => {
    getAlbumsForUser(userId).then(albums => {
      albums.forEach(async album => {
        let photos = await getPhotosForAlbum(album.id);
        album.photoCount = photos.length;
        this.albumPhotoListMapping[album.id] = photos;
        this.rawAlbumList.push(album);
      });
    });
  };

  @action
  setCurrentAlbum = album => {
    this.currentAlbum = album;
  };

  @action
  setCurrentPhoto = photo => {
    this.currentPhoto = photo;
  };

  @action
  clearData() {
    this.rawAlbumList = [];
    this.albumPhotoListMapping = new Map();
    this.currentAlbum = null;
    this.currentPhoto = null;
  }

  @computed
  get albumList() {
    let filteredArray = toJS(this.rawAlbumList)
      .concat()
      .sort((lhsUser, rhsUser) => {
        var idLeft = lhsUser.id;
        var idRight = rhsUser.id; // ignore upper and lowercase
        if (idLeft < idRight) {
          return -1;
        }
        if (idLeft > idRight) {
          return 1;
        }
        // names must be equal
        return 0;
      });
    return filteredArray;
  }
}
const photoStore = new PhotoStore();
export default photoStore;
