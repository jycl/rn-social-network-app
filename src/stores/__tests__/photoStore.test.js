import photoStore from "../photoStore";

describe("PhotoStore tests ", () => {
  it("Snapshot test of PhotoStore", () => {
    expect(photoStore).toMatchSnapshot();
  });

  it("Check state values initially and after changes", () => {
    expect(photoStore.rawAlbumList).toHaveLength(0);
    const albumList = [
      {
        id: 7,
        title: "July"
      },
      {
        id: 3,
        title: "March"
      },
      {
        id: 4,
        title: "April"
      },
      {
        id: 1,
        title: "January"
      }
    ];
    //set up mock function to load user list and set to rawUserList
    photoStore.loadPhotoAlbums = jest.fn().mockImplementation(() => {
      photoStore.rawAlbumList = albumList;
    });
    //invoke method and test if raw and filtered list contain correct values
    photoStore.loadPhotoAlbums("1");
    expect(photoStore.rawAlbumList).toHaveLength(4);
    expect(photoStore.albumList).toHaveLength(4);
    expect(photoStore.albumList[0].title).toBe("January");

    photoStore.clearData();
    expect(photoStore.rawAlbumList).toHaveLength(0);
  });

  it("Test album photo mapping", async () => {
    expect(photoStore.rawAlbumList).toHaveLength(0);
    const photoList = [
      {
        id: 4,
        thumbnailUrl: "/kittens"
      },
      {
        id: 1,
        thumbnailUrl: "/puppies"
      }
    ];
    photoStore.getPhotosForAlbum = jest.fn().mockImplementation(albumId => {
      return Promise.resolve(photoList);
    });
    photoStore.loadPhotoAlbums = jest.fn().mockImplementation(async () => {
      let photos = await photoStore.getPhotosForAlbum(1);
      photoStore.albumPhotoListMapping[1] = photos;
    });
    //invoke method and test if raw and filtered list contain correct values
    await photoStore.loadPhotoAlbums("1");
    expect(photoStore.albumPhotoListMapping[1]).toBeTruthy();
  });
});
