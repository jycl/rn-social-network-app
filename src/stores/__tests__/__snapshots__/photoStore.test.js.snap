// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoStore tests  Snapshot test of PhotoStore 1`] = `PhotoStore {}`;
